### PR TITLE
[4.0] RavenDB-11354 High CPU after hitting leader stepdown several times

### PR DESCRIPTION
**Pull request by karmeli87**
The issue is related to code for sending topology change notification from the leader.
We waited to leave 'Follower' state in order to reconnect to a new leader, but it could happened that the leader was changed without leaving the follower state, and so we were connected to the wrong node.
This could caused a circular connection between the nodes which sent the same notification again and again.

### CODE REVIEW

File: src/Raven.Server/NotificationCenter/Notifications/Server/ClusterTopologyChanged.cs [hunk 1]
```diff
@@ -53,7 +53,8 @@ public override DynamicJsonValue ToJson()
                 NodeTag = nodeTag,
                 Status = status,
                 CurrentTerm = term,
-                NodeLicenseDetails = nodeLicenseDetails
+                NodeLicenseDetails = nodeLicenseDetails,
+                IsPersistent = true
```
Comment (ppekrol) [inline src/Raven.Server/NotificationCenter/Notifications/Server/ClusterTopologyChanged.cs:57]: Why do we want to persist this?

File: src/Raven.Server/NotificationCenter/NotificationsBase.cs [hunk 2]
```diff
@@ -17,18 +17,30 @@ public abstract class NotificationsBase : IDisposable
         private TaskCompletionSource<object> _newWebSocket = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
         private TaskCompletionSource<object> _allWebSocketsRemoved = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
 
-        public Task WaitForAllRemoved => _allWebSocketsRemoved.Task;
+        public Task WaitForRemoveAllWebSocketClients
+        {
+            get
+            {
+                lock (_watchersLock)
```
Comment (ayende) [inline src/Raven.Server/NotificationCenter/NotificationsBase.cs]: Can we avoid this lock?

File: src/Raven.Server/NotificationCenter/Notifications/Server/ClusterTopologyChanged.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/NotificationCenter/Notifications/Server/ClusterTopologyChanged.cs:57]: 1. We need to update the topology after reelection even if some nodes are connected after the initial message was sent.
2. Anyway we keep only the recent topology. (the Id defined by the Type in this case)

File: src/Raven.Server/NotificationCenter/NotificationsBase.cs [hunk 3]
```diff
@@ -17,18 +17,30 @@ public abstract class NotificationsBase : IDisposable
...
+        private volatile int _websocketClients;
 
-        public Task WaitForNew()
+        public Task WaitForAnyWebSocketClient
         {
-            lock (_watchersLock)
+            get
             {
-                return _websocketClients > 0 ? Task.CompletedTask : _newWebSocket.Task;
+                lock (_watchersLock)
```
Comment (ayende) [inline src/Raven.Server/NotificationCenter/NotificationsBase.cs]: can we avoid this lock?

Comment (ayende) [reply src/Raven.Server/NotificationCenter/NotificationsBase.cs]: We can use `Exchange` to do this atomically, instead

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 4]
```diff
@@ -255,7 +255,7 @@ private set
 
         public event EventHandler LeaderElected;
 
-        private string _tag;
+        private volatile string _tag;
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: Under what scenario is this desirable?

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 5]
```diff
@@ -484,6 +484,21 @@ private void SwitchToSingleLeader(TransactionOperationContext context)
...
 
+        public async Task WaitForLeaderChange(CancellationToken cts)
+        {
+            var currentLeader = LeaderTag;
+            while (cts.IsCancellationRequested == false)
+            {
+                // we setup the wait _before_ checking the state
+                var task = _stateChanged.Task;
+
+                if (currentLeader != LeaderTag)
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: Better to have the hit here, using `Volatile.Read`, than everywhere.

File: src/Raven.Server/NotificationCenter/NotificationsBase.cs [hunk 6]
```diff
@@ -69,7 +68,7 @@ public IDisposable TrackActions(AsyncQueue<DynamicJsonValue> notificationsQueue,
 
                 if (watcher.Writer is NotificationCenterWebSocketWriter)
                 {
-                    if (_websocketClients == 0)
+                    if (Interlocked.CompareExchange(ref _haveClients, 1, 0) == 0)
```
Comment (ayende) [inline src/Raven.Server/NotificationCenter/NotificationsBase.cs]: You have the wrong order here

**Comment by ayende** [comment]
See how this behaves https://github.com/ravendb/ravendb/pull/6614/commits/c87cbc54d56672ae7d75c9b3626edf20f6a77d61

**Pull request merged**